File: packages/sdk-dart/lib/src/resources/sip_trunks.dart
import '../models/telephony.dart';
import '../page.dart';
import '../transport.dart';

/// SIP trunks: your own carrier, connected.
class SipTrunks {
  /// Binds the resource to a transport.
  const SipTrunks(this._transport);

  final Transport _transport;

  /// Lists trunks. Each carries `platformIp` (allowlist it for outbound)
  /// and `originationUri` (point the carrier at it for inbound).
  Future<Page<SipTrunk>> list(
          {int? limit, String? startingAfter, String? endingBefore}) =>
      _transport.list('/v1/sip-trunks', SipTrunk.fromJson, query: {
        'limit': limit,
        'starting_after': startingAfter,
        'ending_before': endingBefore,
      });

  /// Connects a carrier. The password is stored encrypted and never
  /// returned.
  Future<SipTrunk> create(CreateSipTrunk body) =>
      _transport.request('POST', '/v1/sip-trunks', SipTrunk.fromJson,
          body: body.toJson());

  /// Fetches one trunk.
  Future<SipTrunk> retrieve(String id) => _transport.request(
      'GET', '/v1/sip-trunks/${Uri.encodeComponent(id)}', SipTrunk.fromJson);

  /// Changes a trunk. Only the fields you set are changed.
  Future<SipTrunk> update(String id, UpdateSipTrunk body) => _transport.request(
      'PATCH', '/v1/sip-trunks/${Uri.encodeComponent(id)}', SipTrunk.fromJson,
      body: body.toJson());

  /// Deletes a trunk. Numbers on it can no longer make or take calls.
  Future<void> delete(String id) => _transport.requestVoid(
      'DELETE', '/v1/sip-trunks/${Uri.encodeComponent(id)}');

  /// Asterisk's own view and a live probe, side by side. Free and
  /// read-only; start here when calls fail.
  Future<SipTrunkStatus> status(String id) => _transport.request(
      'GET',
      '/v1/sip-trunks/${Uri.encodeComponent(id)}/status',
      SipTrunkStatus.fromJson);

  /// Recent SIP events. Poll with [sinceId] set to the previous `lastId`.
  Future<SipLogList> logs(String id, {int? sinceId, int? limit}) =>
      _transport.request(
        'GET',
        '/v1/sip-trunks/${Uri.encodeComponent(id)}/logs',
        SipLogList.fromJson,
        query: {'since_id': sinceId, 'limit': limit},
      );

  /// Sends an unauthenticated probe to the carrier. A negative result is
  /// not proof the trunk is broken.
  Future<SipTrunkTest> test(String id) => _transport.request('POST',
      '/v1/sip-trunks/${Uri.encodeComponent(id)}/test', SipTrunkTest.fromJson);
}

File: packages/sdk-ts/src/resources/sip-trunks.ts
import type { Http, RequestOptions } from '../core/http.js';
import type { Page } from '../core/page.js';
import type { CreateSipTrunk, PaginationQuery, SipLogList, SipTrunk, SipTrunkLogsQuery, SipTrunkStatus, SipTrunkTest, UpdateSipTrunk } from '../types.js';

const enc = encodeURIComponent;

/** SIP trunks: your own carrier, connected. */
export class SipTrunks {
    constructor(private readonly http: Http) {}

    list(query: PaginationQuery = {}, opts?: RequestOptions): Promise<Page<SipTrunk>> {
        return this.http.list<SipTrunk>('/v1/sip-trunks', query, opts);
    }

    /**
     * Connect a carrier. The password is stored encrypted and never returned.
     * Allowlist the returned `platform_ip` with the carrier for outbound, and
     * point the carrier at `origination_uri` for inbound.
     */
    create(body: CreateSipTrunk, opts?: RequestOptions): Promise<SipTrunk> {
        return this.http.request<SipTrunk>('POST', '/v1/sip-trunks', { ...opts, body });
    }

    retrieve(id: string, opts?: RequestOptions): Promise<SipTrunk> {
        return this.http.request<SipTrunk>('GET', `/v1/sip-trunks/${enc(id)}`, opts);
    }

    /** Only the fields sent are changed. */
    update(id: string, body: UpdateSipTrunk, opts?: RequestOptions): Promise<SipTrunk> {
        return this.http.request<SipTrunk>('PATCH', `/v1/sip-trunks/${enc(id)}`, { ...opts, body });
    }

    delete(id: string, opts?: RequestOptions): Promise<void> {
        return this.http.request<void>('DELETE', `/v1/sip-trunks/${enc(id)}`, opts);
    }

    /** Asterisk's own view and a live probe, side by side. Free and read-only; start here when calls fail. */
    status(id: string, opts?: RequestOptions): Promise<SipTrunkStatus> {
        return this.http.request<SipTrunkStatus>('GET', `/v1/sip-trunks/${enc(id)}/status`, opts);
    }

    /** Recent SIP events from an in-memory ring buffer. Poll with `since_id` set to the previous `last_id`. */
    logs(id: string, query: SipTrunkLogsQuery = {}, opts?: RequestOptions): Promise<SipLogList> {
        return this.http.request<SipLogList>('GET', `/v1/sip-trunks/${enc(id)}/logs`, { ...opts, query });
    }

    /** Send an unauthenticated probe to the carrier. A negative result is not proof the trunk is broken. */
    test(id: string, opts?: RequestOptions): Promise<SipTrunkTest> {
        return this.http.request<SipTrunkTest>('POST', `/v1/sip-trunks/${enc(id)}/test`, opts);
    }
}
